nemo/collections/nlp/models/intent_slot_classification_model.py

```diff
@@ -0,0 +1,266 @@
+
+        tensorboard_logs = {'train_loss': train_loss, 'lr': self._optimizer.param_groups[0]['lr']}
+        return {'loss': train_loss, 'log': tensorboard_logs}
+
+    def validation_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the validation loop with the data from the validation dataloader
+        passed in as `batch`.
+        """
+        input_ids, input_type_ids, input_mask, loss_mask, subtokens_mask, intent_labels, slot_labels = batch
+        intent_logits, slot_logits = self(
+            input_ids=input_ids, token_type_ids=input_type_ids, attention_mask=input_mask
+        )
+
+        # calculate combined loss for intents and slots
+        intent_loss = self.intent_loss(logits=intent_logits, labels=intent_labels)
+        slot_loss = self.slot_loss(logits=slot_logits, labels=slot_labels, loss_mask=loss_mask)
+        val_loss = self.total_loss(loss_1=intent_loss, loss_2=slot_loss)
+
+        # calculate accuracy metrics for intents and slot reporting
+        # intents
+        preds = torch.argmax(intent_logits, axis=-1)
+        intent_tp, intent_fp, intent_fn = self.intent_classification_report(preds, intent_labels)
+        # slots
+        subtokens_mask = subtokens_mask > 0.5
+        preds = torch.argmax(slot_logits, axis=-1)[subtokens_mask]
+        slot_labels = slot_labels[subtokens_mask]
+        slot_tp, slot_fp, slot_fn = self.slot_classification_report(preds, slot_labels)
+
+        tensorboard_logs = {
+            'val_loss': val_loss,
+            'intent_tp': intent_tp,
+            'intent_fn': intent_fn,
+            'intent_fp': intent_fp,
+            'slot_tp': slot_tp,
+            'slot_fn': slot_fn,
+            'slot_fp': slot_fp,
+        }
+
+        return {'val_loss': val_loss, 'log': tensorboard_logs}
+
+    def validation_epoch_end(self, outputs):
+        """
+        Called at the end of validation to aggregate outputs.
+        :param outputs: list of individual outputs of each validation step.
+        """
+        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
+
+        # calculate metrics and log classification report (separately for intents and slots)
+        tp = torch.sum(torch.stack([x['log']['intent_tp'] for x in outputs]), 0)
+        fn = torch.sum(torch.stack([x['log']['intent_fn'] for x in outputs]), 0)
+        fp = torch.sum(torch.stack([x['log']['intent_fp'] for x in outputs]), 0)
+        intent_precision, intent_recall, intent_f1 = self.intent_classification_report.get_precision_recall_f1(
+            tp, fn, fp, mode='micro'
+        )
+
+        tp = torch.sum(torch.stack([x['log']['slot_tp'] for x in outputs]), 0)
+        fn = torch.sum(torch.stack([x['log']['slot_fn'] for x in outputs]), 0)
+        fp = torch.sum(torch.stack([x['log']['slot_fp'] for x in outputs]), 0)
+        slot_precision, slot_recall, slot_f1 = self.slot_classification_report.get_precision_recall_f1(
+            tp, fn, fp, mode='macro'
```

Comment:
Why one of them is micro and the other macro?

Reply:
I copied first from your code and second from Evelina code. For Intents it is "Micro" and for Slots it is "Macro". I am actually do not know why.

Reply:
Actually it only influence Logging in TensorBoard and probably "micro" is the right one in both cases. I changed it. But then it is different in @ekmb code for Token Classification.